apps/barista-design-system/src/shared/services/scroll-spy.service.ts

```diff
@@ -14,81 +14,113 @@
- * scroll position.
- */
-export class BaScrollSpiedElement implements BaScrollItem {
-  top = 0;
+@Injectable()
+export class BaScrollSpyService {
+  /** Array of spyable elements with corresponding top value */
+  private _spiedElements: BaScrollItem[] = [];
+  /** Unsubscribes from resize and scroll events */
+  private _onStopListening = new Subject<void>();
+  /** Content height after resize and scroll events */
+  private _lastContentHeight: number;
+  /** Last possible scroll top position  */
+  private _lastMaxScrollTop: number;
+  /** Top offset constant  */
+  private _TOPOFFSET = 66;
+  /** Currently active scroll item */
+  activeScrollItem: ReplaySubject<BaScrollItem | null> = new ReplaySubject(1);
 
-  constructor(readonly element: Element, readonly index: number) {}
+  constructor(
+    private _platform: Platform,
+    @Inject(DOCUMENT) private _document: any,
+  ) {}
 
-  /*
-   * Calculate the `top` value, i.e. the value of the `scrollTop` property at which this element
-   * becomes active. The current implementation assumes that window is the scroll-container.
+  /**
+   * @internal Start tracking a group of elements and emitting active elements; i.e. elements that are
+   * currently visible in the viewport. If there was no other group being spied, start listening for
+   * `resize` and `scroll` events.
    */
-  calculateTop(scrollTop: number, topOffset: number): void {
-    this.top = scrollTop + this.element.getBoundingClientRect().top - topOffset;
+  spyOn(elements: Element[]): Observable<BaScrollItem | null> {
+    this._mapToSpiedElementsFrom(elements);
+    fromEvent(window, 'resize')
+      .pipe(auditTime(300), takeUntil(this._onStopListening))
+      .subscribe(() => {
+        this._onResize();
+      });
+    fromEvent(window, 'scroll')
+      .pipe(auditTime(10), takeUntil(this._onStopListening))
+      .subscribe(() => {
+        this._onScrollEvent();
+      });
+    this._onResize();
+    this._spiedElements.forEach((_) => {
+      this._onScroll(this._getScrollTop(), this._lastMaxScrollTop);
+    });
+
+    return this.activeScrollItem.asObservable().pipe(distinctUntilChanged());
   }
-}
 
-/*
- * Represents a group of "scroll-spied" elements. Contains info and methods for efficiently
- * determining which element should be considered "active", i.e. which element has been scrolled
- * passed the top of the viewport.
- */
-export class BaScrollSpiedElementGroup {
-  activeScrollItem: ReplaySubject<BaScrollItem | null> = new ReplaySubject(1);
-  private _spiedElements: BaScrollSpiedElement[];
+  /**
+   * Stop tracking this group of elements and emitting active elements. If there is no other group
+   * being spied, stop listening for `resize` or `scroll` events.
+   */
+  unspy(): void {
+    this.activeScrollItem.complete();
+    this._spiedElements = [];
+    if (!this._spiedElements.length) {
+      this._onStopListening.next();
+    }
```

Comment:
This does not make a lot of sense. You reset the spiedElements to an array with length 0 and in line 86 check if the length is not 0. 
You should always next stop listening, as you do not differentiate if there are elements when you subscribe in lines 62 and 67